htsget-search/src/htsget/vcf_search.rs

```diff
@@ -0,0 +1,407 @@
+  index::{optimize_chunks, Chunk},
+  VirtualPosition,
+};
+use noodles_tabix::{self as tabix};
+use noodles_vcf::{self as vcf};
+
+use crate::{
+  htsget::{Class, Format, HtsGetError, Query, Response, Result, Url},
+  storage::{BytesRange, GetOptions, Storage, UrlOptions},
+};
+
+// TODO: This trait is clearly common across, **at least**, VCF and BAM
+trait VirtualPositionExt {
+  const MAX_BLOCK_SIZE: u64 = 65536;
+
+  /// Get the starting bytes for a compressed BGZF block.
+  fn bytes_range_start(&self) -> u64;
+  /// Get the ending bytes for a compressed BGZF block.
+  fn bytes_range_end(&self, reader: &mut vcf::Reader<bgzf::Reader<File>>) -> u64;
+  fn to_string(&self) -> String;
+}
+
+impl VirtualPositionExt for VirtualPosition {
+  /// This is just an alias to compressed. Kept for consistency.
+  fn bytes_range_start(&self) -> u64 {
+    self.compressed()
+  }
+  /// The compressed part refers always to the beginning of a BGZF block.
+  /// But when we need to translate it into a byte range, we need to make sure
+  /// the reads falling inside that block are also included, which requires to know
+  /// where that block ends, which is not trivial nor possible for the last block.
+  /// The solution used here goes through reading the records starting at the compressed
+  /// virtual offset (coffset) of the end position (remember this will always be the
+  /// start of a BGZF block). If we read the records pointed by that coffset until we
+  /// reach a different coffset, we can find out where the current block ends.
+  /// Therefore this can be used to only add the required bytes in the query results.
+  /// If for some reason we can't read correctly the records we fall back
+  /// to adding the maximum BGZF block size.
+  fn bytes_range_end(&self, reader: &mut vcf::Reader<bgzf::Reader<File>>) -> u64 {
+    if self.uncompressed() == 0 {
+      // If the uncompressed part is exactly zero, we don't need the next block
+      return self.compressed();
+    }
+    get_next_block_position(*self, reader).unwrap_or(self.compressed() + Self::MAX_BLOCK_SIZE)
+  }
+
+  fn to_string(&self) -> String {
+    format!("{}/{}", self.compressed(), self.uncompressed())
+  }
+}
+
+fn get_next_block_position(
+  block_position: VirtualPosition,
+  reader: &mut vcf::Reader<bgzf::Reader<File>>,
+) -> Option<u64> {
+  reader.seek(block_position).ok()?.compressed();
+  let next_block_index = loop {
+    let bytes_read = reader.read_record(&mut String::new()).ok()?;
+    let actual_block_index = reader.virtual_position().compressed();
+    if bytes_read == 0 || actual_block_index > block_position.compressed() {
+      break actual_block_index;
+    }
+  };
+  Some(next_block_index)
+}
+
+pub(crate) struct VcfSearch<'a, S> {
+  storage: &'a S,
+}
+
+impl<'a, S> VcfSearch<'a, S>
+where
+  S: Storage + 'a,
+{
+  const MIN_SEQ_POSITION: i32 = 1; // 1-based
+  const MAX_SEQ_POSITION: i32 = 37450; // see https://github.com/zaeleus/noodles/issues/25#issuecomment-868871298
```

Comment:
```suggestion
  const MAX_SEQ_POSITION: i32 = (1 << 29) - 1; // see https://github.com/zaeleus/noodles/issues/25#issuecomment-868871298
```

The max supported position is 2<sup>29</sup> - 1. The max bin ID is 37450.

Reply:
I mixed them up 😥. Thanks so much!

Reply:
@zaeleus Thanks so much for chiming in here, very much appreciated! ;)